flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/SharedSlot.java

```diff
@@ -215,10 +218,19 @@ public void release(Throwable cause) {
 			slotContextFuture.isDone(),
 			"Releasing of the shared slot is expected only from its successfully allocated physical slot ({})",
 			physicalSlotRequestId);
-		for (ExecutionVertexID executionVertexId : requestedLogicalSlots.keySetA()) {
-			LOG.debug("Release {}", getLogicalSlotString(executionVertexId));
+		LOG.debug("Release shared slot ({})", physicalSlotRequestId);
+
+		// copy the logical slot collection to avoid ConcurrentModificationException
+		// if logical slot releases cause cancellation of other executions
+		// which will try to call returnLogicalSlot and modify requestedLogicalSlots collection
+		Map<ExecutionVertexID, CompletableFuture<SingleLogicalSlot>> logicalSlotFutures = requestedLogicalSlots
+			.keySetA()
+			.stream()
+			.collect(Collectors.toMap(executionVertexId -> executionVertexId, requestedLogicalSlots::getValueByKeyA));
```

Comment:
Are we only creating this map because of the debug logging? Wouldn't it be simpler to take `requestedLogicalSlots.values()`?

Reply:
True, it would be computationally cheaper but I would prefer to keep the full logging statement with all ids.
I do not think that the slot will be shared by unreasonably high number of executions.

Reply:
What about doing this computation only if the DEBUG log level is activated?